core/src/test/java/com/sequenceiq/cloudbreak/job/stackpatcher/ExistingStackPatcherServiceProviderTest.java

```diff
@@ -0,0 +1,89 @@
+package com.sequenceiq.cloudbreak.job.stackpatcher;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.Mockito.lenient;
+
+import java.lang.reflect.Field;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.util.ReflectionUtils;
+
+import com.sequenceiq.cloudbreak.domain.converter.StackPatchTypeConverter;
+import com.sequenceiq.cloudbreak.domain.stack.StackPatchType;
+import com.sequenceiq.cloudbreak.service.stackpatch.ExistingStackPatchService;
+
+@ExtendWith(MockitoExtension.class)
+class ExistingStackPatcherServiceProviderTest {
+
+    private static final StackPatchType STACK_PATCH_TYPE = StackPatchType.UNBOUND_RESTART;
+
+    @InjectMocks
+    private ExistingStackPatcherServiceProvider underTest;
+
+    @Mock
+    private ExistingStackPatchService existingStackPatchService;
+
+    @Mock
+    private StackPatchTypeConverter stackPatchTypeConverter;
+
+    @BeforeEach
+    void setUp() {
+        lenient().when(existingStackPatchService.getStackPatchType()).thenReturn(STACK_PATCH_TYPE);
+        setStackPatchServices();
+    }
+
+    @Test
+    void shouldFailWithNullName() {
+        String stackPatchTypeName = null;
+        assertThatThrownBy(() -> underTest.provide(stackPatchTypeName))
+                .isInstanceOf(UnknownStackPatchTypeException.class)
+                .hasMessage(String.format("Stack patch type %s is unknown", stackPatchTypeName));
+    }
+
+    @Test
+    void shouldFailWithUnknownName() {
+        String stackPatchTypeName = "UNKNOWN_TYPE";
+        assertThatThrownBy(() -> underTest.provide(stackPatchTypeName))
+                .isInstanceOf(UnknownStackPatchTypeException.class)
+                .hasMessage(String.format("Stack patch type %s is unknown", stackPatchTypeName));
+    }
+
+    @Test
+    void shouldFailWithTypeWithoutService() {
+        StackPatchType stackPatchType = StackPatchType.LOGGING_AGENT_AUTO_RESTART;
+        assertThatThrownBy(() -> underTest.provide(stackPatchType))
+                .isInstanceOf(UnknownStackPatchTypeException.class)
+                .hasMessage("No stack patcher implementation found for type " + stackPatchType);
+    }
+
+    @Test
+    void shouldReturnServiceWithValidType() throws UnknownStackPatchTypeException {
+        ExistingStackPatchService stackPatchService = underTest.provide(STACK_PATCH_TYPE);
+
+        assertThat(stackPatchService).isEqualTo(this.existingStackPatchService);
+    }
+
+    /**
+     * workaround for collection injection
+     */
+    private void setStackPatchServices() {
```

Comment:
Instead of this workaround you can use @Spy annotation on top of a collection implementation so you can add the existingStackPatchService in the before method to the collection.